app/src/main/java/org/thoughtcrime/securesms/ApplicationContext.java

```diff
@@ -0,0 +1,433 @@
+                            .addNonBlocking(EmojiSource::refresh)
+                            .addNonBlocking(() -> ApplicationDependencies.getGiphyMp4Cache().onAppStart(this))
+                            .addPostRender(() -> RateLimitUtil.retryAllRateLimitedMessages(this))
+                            .addPostRender(this::initializeExpiringMessageManager)
+                            .addPostRender(() -> SignalStore.settings().setDefaultSms(Util.isDefaultSmsProvider(this)))
+                            .addPostRender(() -> DownloadLatestEmojiDataJob.scheduleIfNecessary(this))
+                            .addPostRender(EmojiSearchIndexDownloadJob::scheduleIfNecessary)
+                            .addPostRender(() -> DatabaseFactory.getMessageLogDatabase(this).trimOldMessages(System.currentTimeMillis(), FeatureFlags.retryRespondMaxAge()))
+                            .execute();
+
+    Log.d(TAG, "onCreate() took " + (System.currentTimeMillis() - startTime) + " ms");
+    SignalLocalMetrics.ColdStart.onApplicationCreateFinished();
+    Tracer.getInstance().end("Application#onCreate()");
+  }
+
+  @Override
+  public void onForeground() {
+    long startTime = System.currentTimeMillis();
+    Log.i(TAG, "App is now visible.");
+
+    ApplicationDependencies.getFrameRateTracker().start();
+    ApplicationDependencies.getMegaphoneRepository().onAppForegrounded();
+    ApplicationDependencies.getDeadlockDetector().start();
+    SubscriptionKeepAliveJob.launchSubscriberIdKeepAliveJobIfNecessary();
+
+    SignalExecutors.BOUNDED.execute(() -> {
+      FeatureFlags.refreshIfNecessary();
+      ApplicationDependencies.getRecipientCache().warmUp();
+      RetrieveProfileJob.enqueueRoutineFetchIfNecessary(this);
+      GroupV1MigrationJob.enqueueRoutineMigrationsIfNecessary(this);
+      executePendingContactSync();
+      KeyCachingService.onAppForegrounded(this);
+      ApplicationDependencies.getShakeToReport().enable();
+      checkBuildExpiration();
+    });
+
+    Log.d(TAG, "onStart() took " + (System.currentTimeMillis() - startTime) + " ms");
+  }
+
+  @Override
+  public void onBackground() {
+    Log.i(TAG, "App is no longer visible.");
+    KeyCachingService.onAppBackgrounded(this);
+    ApplicationDependencies.getMessageNotifier().clearVisibleThread();
+    ApplicationDependencies.getFrameRateTracker().stop();
+    ApplicationDependencies.getShakeToReport().disable();
+    ApplicationDependencies.getDeadlockDetector().stop();
+  }
+
+  public PersistentLogger getPersistentLogger() {
+    return persistentLogger;
+  }
+
+  public void checkBuildExpiration() {
+    if (Util.getTimeUntilBuildExpiry() <= 0 && !SignalStore.misc().isClientDeprecated()) {
+      Log.w(TAG, "Build expired!");
+      SignalStore.misc().markClientDeprecated();
+    }
+  }
+
+  private void initializeSecurityProvider() {
+    try {
+      Class.forName("org.signal.aesgcmprovider.AesGcmCipher");
+    } catch (ClassNotFoundException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)